### ISSUE TITLE
question: Cannot create pmem-csi volume 

### ISSUE THREAD

**Issue by obedmr** [intel/pmem-CSI#107]
When I tried to create a pmem-csi PVC, it seems like it's not having the capacity to provide it.

My k8s cluster is running with 1 master and 3 workers, only 1 worker has the pmem-csi labeled.

**This is the error I get in the pvc creation:**
```
failed to provision volume with StorageClass "pmem-csi-sc": rpc error: code = Unavailable desc = No node found with 8589934592 capacaity
```

**This is the relevant log from the pod that is running the node with persistence memory enabled:**

```
.
.
.
I1213 17:46:31.993469       1 glog.go:79] total used: 0
I1213 17:46:33.070068       1 glog.go:79] CheckVG: Bus: ndbus0
I1213 17:46:33.070457       1 glog.go:79] Region: region1
I1213 17:46:33.070465       1 glog.go:79] NsMode: fsdax
I1213 17:46:33.071043       1 glog.go:79] No active namespaces in region region1
I1213 17:46:33.071058       1 glog.go:79] NsMode: sector
I1213 17:46:33.071121       1 glog.go:79] No active namespaces in region region1
I1213 17:46:33.071128       1 glog.go:79] Region: region0
I1213 17:46:33.071133       1 glog.go:79] NsMode: fsdax
I1213 17:46:33.071724       1 glog.go:79] No active namespaces in region region0
I1213 17:46:33.071732       1 glog.go:79] NsMode: sector
I1213 17:46:33.071793       1 glog.go:79] No active namespaces in region region0
I1213 17:46:37.382378       1 main.go:138] Version: v0.4.1-0-gb3ef1f69a
```

**Here's the command how my VM is started:**
```
qemu-system-x86_64 -drive if=virtio,file=/data/cow_image.qcow2,format=qcow2 \
 -drive file=/data/seed.iso,if=virtio,format=raw -nodefaults \
-device virtio-balloon-pci,id=balloon0 -realtime mlock=off -msg timestamp=on \
-chardev pty,id=charserial0 -device isa-serial,chardev=charserial0,id=serial0 \
-serial stdio -enable-kvm -machine accel=kvm,usb=off -vga qxl \
-display none -m 4G,slots=4,maxmem=32G -smp 4 -machine pc,accel=kvm,nvdimm=on \
-device virtio-net-pci,netdev=net0,mac=02:42:ac:11:00:09 -netdev tap,id=net0,vhost=on,fd=3 \
-object memory-backend-file,id=mem1,share,mem-path=/tmp/f27nvdimm0,size=4G -device nvdimm,memdev=mem1,id=nv1,label-size=2M \
-object memory-backend-file,id=mem2,share,mem-path=/tmp/f27nvdimm1,size=4G -device nvdimm,memdev=mem2,id=nv2,label-size=2M -vnc unix:/data/vnc
```

I ran this in order to enable the NVIDIMM regions
```
ndctl disable-region all
ndctl init-labels all
ndctl enable-region all
```

Is there something that is missing to run? I saw some documentation about the usage of https://github.com/intel/ipmctl, is it only for the cases when I have the Optane HW?

**Comment by avalluri** [comment]
> When I tried to create a pmem-csi PVC, it seems like it's not having the capacity to provide it.
> 
> My k8s cluster is running with 1 master and 3 workers, only 1 worker has the pmem-csi labeled.
> 
> **This is the error I get in the pvc creation:**
> 
> ```
> failed to provision volume with StorageClass "pmem-csi-sc": rpc error: code = Unavailable desc = No node found with 8589934592 capacaity
> ```
> **This is the relevant log from the pod that is running the node with persistence memory enabled:**
> 
> ```
> .
> .
> .
> I1213 17:46:31.993469       1 glog.go:79] total used: 0
> I1213 17:46:33.070068       1 glog.go:79] CheckVG: Bus: ndbus0
> I1213 17:46:33.070457       1 glog.go:79] Region: region1
> I1213 17:46:33.070465       1 glog.go:79] NsMode: fsdax
> I1213 17:46:33.071043       1 glog.go:79] No active namespaces in region region1
> I1213 17:46:33.071058       1 glog.go:79] NsMode: sector
> I1213 17:46:33.071121       1 glog.go:79] No active namespaces in region region1
> I1213 17:46:33.071128       1 glog.go:79] Region: region0
> I1213 17:46:33.071133       1 glog.go:79] NsMode: fsdax
> I1213 17:46:33.071724       1 glog.go:79] No active namespaces in region region0
> I1213 17:46:33.071732       1 glog.go:79] NsMode: sector
> I1213 17:46:33.071793       1 glog.go:79] No active namespaces in region region0
> I1213 17:46:37.382378       1 main.go:138] Version: v0.4.1-0-gb3ef1f69a
> ```
> **Here's the command how my VM is started:**
> 
> ```
> qemu-system-x86_64 -drive if=virtio,file=/data/cow_image.qcow2,format=qcow2 \
>  -drive file=/data/seed.iso,if=virtio,format=raw -nodefaults \
> -device virtio-balloon-pci,id=balloon0 -realtime mlock=off -msg timestamp=on \
> -chardev pty,id=charserial0 -device isa-serial,chardev=charserial0,id=serial0 \
> -serial stdio -enable-kvm -machine accel=kvm,usb=off -vga qxl \
> -display none -m 4G,slots=4,maxmem=32G -smp 4 -machine pc,accel=kvm,nvdimm=on \
> -device virtio-net-pci,netdev=net0,mac=02:42:ac:11:00:09 -netdev tap,id=net0,vhost=on,fd=3 \
> -object memory-backend-file,id=mem1,share,mem-path=/tmp/f27nvdimm0,size=4G -device nvdimm,memdev=mem1,id=nv1,label-size=2M \
> -object memory-backend-file,id=mem2,share,mem-path=/tmp/f27nvdimm1,size=4G -device nvdimm,memdev=mem2,id=nv2,label-size=2M -vnc unix:/data/vnc
> ```

This s showing that your qemu is configured with two 4GB nvdimms,  so with this configuration you can use max ~4GB volume, Can you please try creating PVC with less than 4GB.

> Is there something that is missing to run? I saw some documentation about the usage of https://github.com/intel/ipmctl, is it only for the cases when I have the Optane HW?

Yes, ipmctl tool is to initialize physical dimms.

**Comment by okartau** [comment]
I think before that this scenario blocks by the fact that default namespace size id 32GB, means the drivers fails to create even first namespace. In emulated case, it makes sense to use smaller than default namespace size using pmem-csi driver added option, for example `-namespacesize 2`
So that in yaml file this config becomes for exmaple:
```
        image: REGISTRY:5000/pmem-ns-init:canary
        args: [ "-v=5", "-namespacesize=2" ]
```

The 32 GB default size was selected real HW  in mind where device will be like 128 G and more. What is the good/optimal value is bit open, also some thoughts in issue #58.
In short, your emulated NVDIMM size and value given to -namespacesize should result in balance
such as at least one, better few namespaces can be create inside one NVDIMM.
Note also there is some overhead so with NVDIMM=8G and namespacesize=4G you will get one, not two, and with such round numbers the unused overhead is close to 4G

**Comment by okartau** [comment]
another hint, use ndctl on host to list the actual namespaces state, it's bit cleaner to examine it there
compared to pmem-csi driver logs

**Comment by obedmr** [comment]
@avalluri and @okartau thanks for the help, I created a volume of 1Gb. 
As far I can see, my regions seem to have enough room to create a 1G volume.

```
ndctl list -R
[
  {
    "dev":"region1",
    "size":4292870144,
    "available_size":4292870144,
    "type":"pmem",
    "numa_node":0,
    "iset_id":10248195696374871,
    "persistence_domain":"unknown"
  },
  {
    "dev":"region0",
    "size":4292870144,
    "available_size":4292870144,
    "type":"pmem",
    "numa_node":0,
    "iset_id":10248187106440278,
    "persistence_domain":"unknown"
  }
]
```

I'm still getting this error:
```
I1213 23:22:06.862312       1 glog.go:79] Enabling volume access mode: SINGLE_NODE_WRITER
I1213 23:22:06.862842       1 glog.go:79] NewPmemDeviceManagerLVM: Bus: ndbus0
I1213 23:22:06.863214       1 glog.go:79] NewPmemDeviceManagerLVM: Region: region1
I1213 23:22:06.863230       1 glog.go:58] Executing: vgs ndbus0region1fsdax
I1213 23:22:06.918079       1 glog.go:58] Output:   Volume group "ndbus0region1fsdax" not found
  Cannot process volume group ndbus0region1fsdax
I1213 23:22:06.918145       1 glog.go:79] NewPmemDeviceManagerLVM: VG ndbus0region1fsdax non-existent, skip
I1213 23:22:06.918202       1 glog.go:58] Executing: vgs ndbus0region1sector
I1213 23:22:06.959001       1 glog.go:58] Output:   Volume group "ndbus0region1sector" not found
  Cannot process volume group ndbus0region1sector
I1213 23:22:06.959053       1 glog.go:79] NewPmemDeviceManagerLVM: VG ndbus0region1sector non-existent, skip
I1213 23:22:06.959101       1 glog.go:79] NewPmemDeviceManagerLVM: Region: region0
I1213 23:22:06.959130       1 glog.go:58] Executing: vgs ndbus0region0fsdax
I1213 23:22:07.031182       1 glog.go:58] Output:   Volume group "ndbus0region0fsdax" not found
  Cannot process volume group ndbus0region0fsdax
I1213 23:22:07.031259       1 glog.go:79] NewPmemDeviceManagerLVM: VG ndbus0region0fsdax non-existent, skip
I1213 23:22:07.031298       1 glog.go:58] Executing: vgs ndbus0region0sector
I1213 23:22:07.099042       1 glog.go:58] Output:   Volume group "ndbus0region0sector" not found
  Cannot process volume group ndbus0region0sector
I1213 23:22:07.099150       1 glog.go:79] NewPmemDeviceManagerLVM: VG ndbus0region0sector non-existent, skip
I1213 23:22:07.099255       1 glog.go:79] Enabling node service capability: STAGE_UNSTAGE_VOLUME
I1213 23:22:07.099602       1 glog.go:79] Enabling controller service capability: PUBLISH_UNPUBLISH_VOLUME
I1213 23:22:07.101561       1 glog.go:79] Listening for connections on address: &net.UnixAddr{Name:"/csi/csi.sock", Net:"unix"}
I1213 23:22:07.101953       1 glog.go:58] Connecting to 10.97.108.156:10000
Connecting to Registry at : tcp://10.97.108.156:10000
```

And my VMs setup is:
```
qemu-system-x86_64 -drive if=virtio,file=/data/cow_image.qcow2,format=qcow2 
-drive file=/data/seed.iso,if=virtio,format=raw -nodefaults 
-device virtio-balloon-pci,id=balloon0 -realtime mlock=off -msg timestamp=on 
-chardev pty,id=charserial0 -device isa-serial,chardev=charserial0,id=serial0 
-serial stdio -enable-kvm -machine accel=kvm,usb=off -vga qxl -display none -m 8G,slots=4,maxmem=16G -smp 4 -machine pc,accel=kvm,nvdimm=on 
-device virtio-net-pci,netdev=net0,mac=02:42:ac:11:00:09 -netdev tap,id=net0,vhost=on,fd=3 
-object memory-backend-file,id=mem1,share,mem-path=/tmp/f27nvdimm0,size=4G 
-device nvdimm,memdev=mem1,id=nv1,label-size=2M 
-object memory-backend-file,id=mem2,share,mem-path=/tmp/f27nvdimm1,size=4G -device nvdimm,memdev=mem2,id=nv2,label-size=2M -vnc unix:/data/vnc
```

**Comment by obedmr** [comment]
@okartau I'm actually running it in K8s, I'm following this part of the main README https://github.com/intel/pmem-CSI#run-as-kubernetes-deployment. As far I understand `pmem-ns-init` and `pmem-vgm` are running on init containers in `pmem-csi` containers that are running on each container.

**Comment by okartau** [comment]
hm, re-read of start of thread shows you attempt pvc, so it means you have Kubernetes context amd my guidance about steps like in run_driver is off-topic.
But question remains, whats your deployment scheme, has pmem-vgm been run?
In kubernetes case, we have deployment example documented in README, using manifests in deploy/, please check recently updated content there suitable for k8s  v1.12.
The pmem-csi.yaml has entries taking care of starting all 3 stages.
in k8s logs you can examine the stage logs separately.
pmem-vgm log should show creation (or re-use) of LVM volume groups.
It is also helpful to re-check using `vgs` on host side, may need to use `vgs --foreign` as vg ops were running inside container

**Comment by obedmr** [comment]
Yes, I'm using the k8s v1.12 yml file from `deploy`, here's the full log in the `pod/pmem-csi-xxx`

`kubectl logs pod/pmem-csi-86f7h   --all-containers` 
http://pastebin.intel.com:8080/sijetejose

For some reason, this error keeps appearing:
```
 Executing: vgs ndbus0region0fsdax
I1213 23:54:25.562723       1 glog.go:58] Output:   Volume group "ndbus0region0fsdax" not found
```
I'm not sure if that may be the issue.

**Comment by obedmr** [comment]
Also, I ran the `vgs` command and showing as follows:
```
root@k8s-ubuntu-pmem-worker1:/home/ubuntu# vgs --foreign
root@k8s-ubuntu-pmem-worker1:/home/ubuntu#
```

**Comment by okartau** [comment]
thanks for log, now I start getting some traction of it.
You have 2 regions, both with available_size:4292870144
You have configured 4G namespacesize:
glog.go:79] Configured namespacesize; 4 GB

pmem-ns-init (1st stage) checks each region and tries to create 4294967296 byte namespaces.
But 4294967296-byte namespace does not fit into free space 4292870144,
therefore it can't create any:
`0 fsdax-namespaces of size 4294967296 possible in region region1`

the reason why 4G cant fit into 4G is namespace overhead (I think it's 4 MB).
Please try with reduced namespacesize, like 3G or 2G (you should get 1 namespace created)
or 1G (you should get 3 namespaces created)

**Comment by obedmr** [comment]
@okartau, yeah, that totally makes sense. I'm almost there. I changed the namespace to 2GB and it seems like it now creates at least one namespace. I'm still getting some weird error later. I'm digging more to find the root cause of this.

```
> kubectl log pod/pmem-csi-v22jt   --all-containers -f
log is DEPRECATED and will be removed in a future version. Use logs instead.
I1214 07:14:02.434865       1 glog.go:79] Configured namespacesize; 2 GB
I1214 07:14:02.435347       1 glog.go:79] Create fsdax-namespaces in region1, allowed 100 %:
total       :       4292870144
avail       :       4292870144
can use     :       4292870144
I1214 07:14:02.435882       1 glog.go:79] total used: 0
I1214 07:14:02.435903       1 glog.go:79] 1 fsdax-namespaces of size 2147483648 possible in region region1
I1214 07:14:02.435910       1 glog.go:79] Creating namespace 0
I1214 07:14:02.439541       1 glog.go:79] setting namespace sector size ...: 0
I1214 07:14:02.442212       1 glog.go:79] setting pfn
libndctl: ndctl_pfn_enable: pfn1.0: failed to enable
W1214 07:14:02.443641       1 glog.go:84] Failed to create namespace:pfn: failed to enable
I1214 07:14:02.443672       1 glog.go:79] Create sector-namespaces in region1, allowed 0 %:
total       :       4292870144
avail       :       4292870144
can use     :                0
I1214 07:14:02.443909       1 glog.go:79] total used: 0
I1214 07:14:02.443936       1 glog.go:79] Create fsdax-namespaces in region0, allowed 100 %:
total       :       4292870144
avail       :       4292870144
can use     :       4292870144
I1214 07:14:02.444540       1 glog.go:79] total used: 0
I1214 07:14:02.444558       1 glog.go:79] 1 fsdax-namespaces of size 2147483648 possible in region region0
I1214 07:14:02.444563       1 glog.go:79] Creating namespace 0
I1214 07:14:02.447361       1 glog.go:79] setting namespace sector size ...: 0
I1214 07:14:02.448955       1 glog.go:79] setting pfn
libndctl: ndctl_pfn_enable: pfn0.0: failed to enable
W1214 07:14:02.450249       1 glog.go:84] Failed to create namespace:pfn: failed to enable
I1214 07:14:02.450274       1 glog.go:79] Create sector-namespaces in region0, allowed 0 %:
total       :       4292870144
avail       :       4292870144
can use     :                0
I1214 07:14:02.450429       1 glog.go:79] total used: 0
I1214 07:14:02.747019       1 glog.go:79] CheckVG: Bus: ndbus0
I1214 07:14:02.747613       1 glog.go:79] Region: region1
I1214 07:14:02.747624       1 glog.go:79] NsMode: fsdax
I1214 07:14:02.748471       1 glog.go:79] No active namespaces in region region1
I1214 07:14:02.748491       1 glog.go:79] NsMode: sector
I1214 07:14:02.748572       1 glog.go:79] No active namespaces in region region1
I1214 07:14:02.748581       1 glog.go:79] Region: region0
I1214 07:14:02.748587       1 glog.go:79] NsMode: fsdax
I1214 07:14:02.749262       1 glog.go:79] No active namespaces in region region0
I1214 07:14:02.749271       1 glog.go:79] NsMode: sector
I1214 07:14:02.749361       1 glog.go:79] No active namespaces in region region0
I1214 07:14:09.589779       1 main.go:138] Version: v0.4.1-0-gb3ef1f69
I1214 07:14:09.589934       1 main.go:145] Attempting to open a gRPC connection with: "/csi/csi.sock"
I1214 07:14:09.589967       1 connection.go:68] Connecting to /csi/csi.sock
I1214 07:14:09.591170       1 connection.go:95] Still trying, connection is CONNECTING
I1214 07:14:09.591985       1 connection.go:95] Still trying, connection is TRANSIENT_FAILURE
I1214 07:14:10.591444       1 connection.go:95] Still trying, connection is CONNECTING
I1214 07:14:10.591504       1 connection.go:92] Connected
I1214 07:14:10.591513       1 main.go:153] Calling CSI driver to discover driver name.
I1214 07:14:10.591529       1 connection.go:136] GRPC call: /csi.v0.Identity/GetPluginInfo
I1214 07:14:10.591538       1 connection.go:137] GRPC request:
I1214 07:14:10.592031       1 connection.go:139] GRPC response: name:"pmem-csi" vendor_version:"0.0.1"
I1214 07:14:10.592074       1 connection.go:140] GRPC error: <nil>
I1214 07:14:10.592085       1 main.go:161] CSI driver name: "pmem-csi"
I1214 07:14:10.592092       1 main.go:165] Loading kubeconfig.
I1214 07:14:10.592295       1 node_register.go:55] Calling CSI driver to discover node ID.
I1214 07:14:10.592311       1 connection.go:136] GRPC call: /csi.v0.Node/NodeGetId
I1214 07:14:10.592317       1 connection.go:137] GRPC request:
I1214 07:14:10.592706       1 connection.go:139] GRPC response: node_id:"k8s-ubuntu-pmem-worker1"
I1214 07:14:10.592731       1 connection.go:140] GRPC error: <nil>
I1214 07:14:10.592752       1 node_register.go:63] CSI driver node ID: "k8s-ubuntu-pmem-worker1"
I1214 07:14:10.592803       1 node_register.go:86] Starting Registration Server at: /registration/pmem-csi-reg.sock
I1214 07:14:10.592883       1 node_register.go:93] Registration Server started at: /registration/pmem-csi-reg.sock
I1214 07:14:10.597913       1 main.go:111] Received GetInfo call: &InfoRequest{}
I1214 07:14:10.605026       1 main.go:121] Received NotifyRegistrationStatus call: &RegistrationStatus{PluginRegistered:true,Error:,}
```

When I try to create the pvc I'm still getting:
```
  Warning  ProvisioningFailed    14s (x2 over 14s)  pmem-csi_pmem-csi-controller-0_8d858b22-ff70-11e8-838d-0a580af40303  failed to provision volume with StorageClass "pmem-csi-sc": rpc error: code = Unavailable desc = No node found with 1073741824 capacaity
```

**Comment by okartau** [comment]
Now we hit next issues related to emulated NVDIMM.
`Failed to create namespace:pfn: failed to enable`

We have seen series of those.
Emulated NVDIMM in general works better if you only enable one device.
Having 2 devices was my own idea to get emulated env closer to real HW, and that
got documented in READMEs, but it actually causes some extra trouble.
You can make success chances higher by removing 2nd emulated device,
and use one larger.
I think I need to update README stating the same.

There is BTW recently added code in test/ which automates qemu VMs creation and k8s cluster setup., using ClearLinux targets. In such a system NVDIMM (one device) emulation has been quite 
stable.

**Comment by okartau** [comment]
I created another PR #109 to stop advertising dual NVDIMM in emulation, which seems to create more trouble than gain

**Comment by obedmr** [comment]
@okartau, quick question, do you know if it has been tested in other distros different than ClearLinux?

**Comment by okartau** [comment]
I can tell where we (dev team of 2 so far been active with that) have been running pmem-csi:
VMs with emulated NVDIMM (as described in READMEs): Ubuntu 18.04.1
VMs created by scripts in test/: Clearlinux, various versions
Phys host with real Pmem devices: Clearlinux
Phys host with real Pmem devices: Fedora 28

As we had not systematic testing set up, this has not been automated, but it has been
verified on those via manual run.
As most(all?) functioning happens in containers (if not running stand-alone binaries),
it should not matter too much what is the distro, if it runs Docker

**Comment by okartau** [comment]
namespace creation was improved since this issue was created. 
Is it better now?

**Comment by obedmr** [comment]
@okartau it has been working now in Ubuntu. I'm experiencing some issues with ClearLinux. Have you ever seen thise error:
```
box ➜  pmem-csi git:(devel) ✗  kubectl logs pod/pmem-csi-vc8dm     --all-containers
I0117 16:31:04.535770       1 glog.go:79] Configured namespacesize; 2 GB
I0117 16:31:04.536109       1 glog.go:79] Create fsdax-namespaces in region0, allowed 100 %:
total       :       4294967296
avail       :       4294967296
can use     :       4294967296
I0117 16:31:04.536608       1 glog.go:79] 4294967296 bytes calculated available after scan for existing fsdax-mode namespaces
I0117 16:31:04.536638       1 glog.go:79] Calculated canUse:4294967296, available by Region info:4294967296
I0117 16:31:04.536667       1 glog.go:79] Create next 2147483648-bytes fsdax-namespace
W0117 16:31:04.536925       1 glog.go:84] Failed to create namespace:Failed to set namespace uid: Read-only file system
I0117 16:31:04.536953       1 glog.go:79] Create sector-namespaces in region0, allowed 0 %:
total       :       4294967296
avail       :       4294967296
can use     :                0
```

This is  the VM qemu's command I'm using for both Clear and Ubuntu. Ubuntu works and Clear not.
```
emu-system-x86_64 -drive if=virtio,file=/data/cow_image.qcow2,format=qcow2,id=data \
-drive file=/data/seed.iso,if=virtio,format=raw -nodefaults \
-device virtio-balloon-pci,id=balloon0 -realtime mlock=off -msg timestamp=on \
-chardev pty,id=charserial0 -device isa-serial,chardev=charserial0,id=serial0 \
-serial stdio -object rng-random,filename=/dev/urandom,id=rng0 \
-device virtio-rng-pci,rng=rng0 -enable-kvm -machine accel=kvm,usb=off -vga qxl \
-display none -m 4G,slots=4,maxmem=8G -smp 4 -machine pc,accel=kvm,nvdimm=on \
-device virtio-net-pci,netdev=net0,mac=02:42:ac:11:00:05 \
-netdev tap,id=net0,vhost=on,fd=3 \
-object memory-backend-file,id=mem1,share,mem-path=/data/nvdimm0,size=4295098368 \
-device nvdimm,memdev=mem1,id=nv1,label-size=128K \
-vnc unix:/data/vnc -bios /OVMF.fd
```

**Comment by obedmr** [comment]
I'm digging into this error, still finding out why it's failing.
```
Failed to create namespace:Failed to set namespace uid: Read-only file system
```

**Comment by okartau** [comment]
sounds like read-only mounted /sys filesystem. We have hit this when we tried cri-o instead of Docker

**Comment by obedmr** [comment]
@okartau, yeah, I'm using cri-o, any hint to fix it? or just move it to docker?

**Comment by okartau** [comment]
yes, see also #112 and how TEST_CRI is used under test/, @pohly added option to select docker vs cri-o,
and with Docker there is no such issue.

**Comment by obedmr** [comment]
@okartau, thanks, I'll move to `docker` then for the moment. Will check @pohly's option

**Comment by obedmr** [comment]
It works in ClearLinux too. Thanks @okartau and @pohly, I think we can close this issue now.

**Comment by okartau** [comment]
thanks, closing
